sources/prairie-dog/src/main.rs

```diff
@@ -0,0 +1,431 @@
+            {
+                subcommand = Some(arg)
+            }
+
+            // Other arguments are passed to the subcommand parser
+            _ => subcommand_args.push(arg),
+        }
+    }
+    match subcommand.as_deref() {
+        Some("capture-dump") => Ok((global_args, Subcommand::CaptureDump {})),
+        Some("configure-boot-mount") => Ok((global_args, Subcommand::ConfigureBootMount {})),
+        Some("load-crash-kernel") => Ok((global_args, Subcommand::LoadCrashKernel {})),
+        None => {
+            return error::Usage {
+                message: format!("Missing subcommand"),
+            }
+            .fail()
+        }
+        Some(x) => {
+            return error::Usage {
+                message: format!("Unknown subcommand '{}'", x),
+            }
+            .fail()
+        }
+    }
+}
+
+/// Wrapper around process::Command that adds error checking.
+fn command<I, S>(bin_path: &str, args: I) -> Result<()>
+where
+    I: IntoIterator<Item = S>,
+    S: AsRef<OsStr>,
+{
+    let mut command = Command::new(bin_path);
+    command.args(args);
+    let output = command
+        .output()
+        .context(error::ExecutionFailure { command })?;
+
+    trace!("stdout: {}", String::from_utf8_lossy(&output.stdout));
+    trace!("stderr: {}", String::from_utf8_lossy(&output.stderr));
+
+    ensure!(
+        output.status.success(),
+        error::CommandFailure { bin_path, output }
+    );
+
+    Ok(())
+}
+
+/// Dumps the memory image in `/proc/vmcore`, which is created when the kernel crashes
+fn capture_dump() -> Result<()> {
+    let kdump_file_path = format!("{}/{}", KDUMP_LOGS_PATH, KDUMP_FILE);
+    let dmesg_file_path = format!("{}/{}", KDUMP_LOGS_PATH, DMESG_DUMP_FILE);
+
+    // Delete previous dumps, if they exist
+    if Path::new(&kdump_file_path).exists() {
+        info!("Deleting existing memory dump");
+        fs::remove_file(&kdump_file_path).context(error::RemoveFile {
+            path: &kdump_file_path,
+        })?;
+    }
+
+    if Path::new(&dmesg_file_path).exists() {
+        info!("Deleting existing dmesg dump");
+        fs::remove_file(&dmesg_file_path).context(error::RemoveFile {
+            path: &dmesg_file_path,
+        })?;
+    }
+
+    info!("Generating memory dump");
+    // Extract kdump-compressed dump file, without empty pages, using
+    // zlib compression
+    command(
+        MAKEDUMPFILE_PATH,
+        &[
+            "-c",
+            "--message-level",
+            "4",
+            "-d",
+            "31",
+            "/proc/vmcore",
+            kdump_file_path.as_ref(),
+        ],
+    )?;
+
+    info!("Generating dmesg dump");
+    // --dump-dmesg generates a dump with only dmesg logs
+    command(
+        MAKEDUMPFILE_PATH,
+        &[
+            "--dump-dmesg",
+            "--message-level",
+            "4",
+            "/proc/vmcore",
+            dmesg_file_path.as_ref(),
+        ],
+    )?;
```

Comment:
I'd reverse these operations since dmesg extraction is quick, and core compression is slow. If we get rebooted before we can finish, at least we'd have some of the data.